frame/staking/src/lib.rs

```diff
@@ -1372,14 +1618,176 @@ impl<T: Trait> Module<T> {
 		// available yet.
 		CurrentEraStartSessionIndex::put(0);
 		BondedEras::mutate(|bonded| bonded.push((0, 0)));
-		Self::select_validators().1
+		Self::select_and_update_validators()
+	}
+
+	/// Checks a given solution and if correct and improved, writes it on chain as the queued result
+	/// of the next round. This may be called by both a signed and an unsigned transaction.
+	fn check_and_replace_solution(
+		winners: Vec<T::AccountId>,
+		compact_assignments: CompactAssignments<T::AccountId, ExtendedBalance>,
+		compute: ElectionCompute,
+	) -> Result<(), Error<T>> {
+		// discard early solutions
+		match Self::era_election_status() {
+			ElectionStatus::None => Err(Error::<T>::PhragmenEarlySubmission)?,
+			ElectionStatus::Open(_) => { /* Open and no solutions received. Allowed. */ },
+		}
+
+		// convert into staked.
+		let staked_assignments = compact_assignments.into_staked::<
+			_,
+			_,
+			T::CurrencyToVote,
+		>(Self::slashable_balance_of);
+
+		// build the support map thereof in order to evaluate.
+		// OPTIMIZATION: we could merge this with the `staked_assignments.iter()` below and
+		// iterate only once.
+		let (supports, num_error) = sp_phragmen::build_support_map::<BalanceOf<T>, T::AccountId>(
+			&winners,
+			&staked_assignments,
+		);
+
+		ensure!(num_error == 0, Error::<T>::PhragmenBogusEdge);
+
+		// score the result. Go further only if it is better than what we already have.
+		let submitted_score = sp_phragmen::evaluate_support(&supports);
```

Comment:
How heavy is that call (together with `build_support_map`), aren't the bonding checks cheaper to perform? I think in general the checks should be done in order of their complexity. If we want to reject cheap scores early on (since we assume there is oging to be a lot of them) we can force the transaction to contain `expected_score` and then reject based on that early, but double check that it matches the `computed_score` later on.

Reply:
Great point. I am working with Shawn to benchmark this function before we merge. That should be the end answer here. For now, I can state: 

- I did my best to prevent any unnecessary computation. 
- I tried to do them in such an order to be able to ditch out as soon as possible. 

For the sake of prosperity, I will try and explain the overall process here. See the added inline comments:


convert into staked. This is 100% needed before we can do anything since the compact type is not interpretable. It does not contain all the weights (it omits the last one since it can be inferred) etc. We _might_ some day realize that the bottle-neck is not the block size and it is CPU. Then we can ditch the whole idea of compactAssignment. But I don't think it is okay to assume this now. The assignment can get big. Current kusama is around 100Kb uncompact an 40kb compacted. See my comment about with number from offline-phragmen. This is O(E)
```
let staked_assignments = compact_assignments.into_staked::<...>(Self::slashable_balance_of);
```

As mentioned we can merge this with the above but I haven't done it yet to keep the code cleaner and be able to use the functions. This is also O(E), so it is not increasing the order of complexity here. 
```
// OPTIMIZATION: we could merge this with the `staked_assignments.iter()` below and
// iterate only once.
let (supports, num_error) = sp_phragmen::build_support_map::<BalanceOf<T>, T::AccountId>(
	&winners,
	&staked_assignments,
);

```

So far we only have built the types.

Finding the score is done as well in O(E).
```
let submitted_score = sp_phragmen::evaluate_support(&supports);
```

As mentioned, we can optimise this a bit by first reading the score from storage.  
if let Some(ElectionResult::<T::AccountId, BalanceOf<T>> {
	score,
	..
}) = Self::queued_elected() {
	// OPTIMIZATION: we can read first the score and then the rest to do less decoding.
	// if the local score is better in any of the three parameters.
	ensure!(
		offchain_election::is_score_better(score, submitted_score),
		Error::<T>::PhragmenWeakSubmission,
	)
}
```

So I generally tried to not look into the storage as much as possible. So far
- We re-built the solution. 
- While building the support, we do have to make sure that all of them have a non-zero stake
- and we checked the score. 

Hence forth we assume: 
- this solution has a valid format and seems to be better. Now we check some stuff that need more on-chain query. 

All of the winners need be validator candidates, as such:
```
// check all winners being actual validators.
for w in winners.iter() {
	ensure!(
		<Validators<T>>::exists(&w),
		Error::<T>::PhragmenBogusWinner,
	)
}
```

check all nominators being bonded, and actually including the claimed vote, and summing up to their ledger stake, as such: 
```
for StakedAssignment { who, distribution } in staked_assignments.iter() {
	... 
}
```

This final note is also very important: 

```
// Note that we don't need to check again ^^ if a particular target in a nomination was
// actually a candidate. we don't explicitly check this but instead do it indirectly.
// build_support_map tells if any target was not in the winners. Also, we check that all
// the winners were actually candidates. Hence, all of the claimed votes are actually
// voting for valid candidates. All we have to check is if they actually come from the
// claimed nominator or not.
```

I still highly appreciate any critic toward this function. 

More important than the complexity is the soundness as well. I tried to cover all cases where a a solution can be invalid, but maybe I have missed some? Just imagine yourself as an adversary, and all the ways that you would try to trick this function. I tried to include all of them in a group of test cases starting with `invalid_phragmen_result_blah_blah`.

Finally, I have placed main doc of this function, for better visibility, in the doc of `submit_election_solution` in `decl_module!`.